power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/dataset.hpp

```diff
@@ -36,16 +38,125 @@ static_assert(is_indptr_mutable_v<writable_dataset_t>);
+            for (auto const& attribute_buffer : attribute_buffers_) {
+                assert(attribute_buffer.meta_attribute != nullptr);
+                auto const& meta_attribute = *attribute_buffer.meta_attribute;
+                ctype_func_selector(
+                    meta_attribute.ctype, [&result, &attribute_buffer, &meta_attribute, this]<typename AttributeType> {
+                        AttributeType const* buffer_ptr =
+                            reinterpret_cast<AttributeType const*>(attribute_buffer.data) + idx_;
+                        AttributeType& attribute_ref =
+                            meta_attribute.template get_attribute<AttributeType>(reinterpret_cast<RawDataPtr>(&result));
+                        attribute_ref = *buffer_ptr;
+                    });
+            }
+            return result;
+        }
+
+      private:
+        friend class ColumnarAttributeRange;
+        friend class iterator;
+
+        Idx idx_{};
+        std::span<AttributeBuffer<Data> const> attribute_buffers_{};
+    };
+
+    class iterator : public boost::iterator_facade<iterator, T, boost::random_access_traversal_tag, Proxy, Idx> {
+      public:
+        using value_type = Proxy;
+
+        iterator() = default;
+        iterator(Idx idx, std::span<AttributeBuffer<Data> const> attribute_buffers)
+            : current_{idx, attribute_buffers} {}
+
+      private:
+        friend class boost::iterator_core_access;
+
+        constexpr auto dereference() const { return current_; }
+        constexpr auto equal(iterator const& other) const { return current_.idx_ == other.current_.idx_; }
+        constexpr auto distance_to(iterator const& other) const { return other.current_.idx_ - current_.idx_; }
+        constexpr void increment() { ++current_.idx_; }
+        constexpr void decrement() { --current_.idx_; }
+        constexpr void advance(Idx n) { current_.idx_ += n; }
+
+        Proxy current_;
+    };
+
+    ColumnarAttributeRange() = default;
+    ColumnarAttributeRange(Idx size, std::vector<AttributeBuffer<Data>> attribute_buffers)
+        : size_{size}, attribute_buffers_{std::move(attribute_buffers)} {}
+    ColumnarAttributeRange(ColumnarAttributeRange::iterator begin, ColumnarAttributeRange::iterator end)
+        : size_{std::distance(begin, end)},
+          attribute_buffers_{begin->attribute_buffers_.begin(), begin->attribute_buffers_.end()} {
+        assert(begin + std::distance(begin, end) == end);
+    }
+
+    constexpr Idx size() const { return size_; }
+    constexpr bool empty() const { return size_ == 0; }
+    iterator begin() const { return get(0); }
+    iterator end() const { return get(size_); }
+    auto iter() const { return boost::make_iterator_range(begin(), end()); }
+    auto operator[](Idx idx) const { return *get(idx); }
+
+  private:
+    iterator get(Idx idx) const { return iterator{idx, attribute_buffers_}; }
+
+    Idx size_{};
+    std::vector<AttributeBuffer<Data>> attribute_buffers_;
```

Comment:
~i wonder whether this should be a `span` as well... really unsure about that tho, but it does make the iterator nicer, i guess~ EDIT: see my other comment on the bug with the current implementation

Reply:
You were right, only `span` is needed now.